### ISSUE TITLE
Conditional Map() Methods

### ISSUE THREAD

**Issue by nathanila** [amaembo/streamex#106]
I've run into this situation a couple of times.

StreamEx.
   ...
   map(item -> condition(item) ? doSomething(item) : doSomethingElse(item)).
   ...

Many times doSomethingElse() is simply "item".

Can you add methods such as the following?

map(Predicate<? super T> condition, Function<? super T, ? extends T> mapper);
map(Predicate<? super T> condition, Function<? super T, ? extends T> positive, Function<? super T, ? extends T> negative);

**Comment by nbardiuk** [comment]
Without library change it could be achieved by a helper function

    static <T, R> Function<T, R> when(Predicate<T> condition, Function<T, R> positive, Function<T, R> negative) {
        return item -> condition.test(item) ? positive.apply(item) : negative.apply(item);
    }

    static <T> Function<T, T> when(Predicate<T> condition, Function<T, T> positive) {
        return item -> condition.test(item) ? positive.apply(item) : item;
    }

    stream.map(when(this::condition, this::then, this::otherwise))
    
    stream.map(when(this::condition, this::update))

It makes possible even such extreme use case as

    stream.map(when(this::condition, this::then, 
               when(this::otherCondition, this::doOther, 
               when(this::anotherCondition, this::doAnother, 
               this::otherwise))))

**Comment by amaembo** [comment]
Ah, you also requested the same for `map`. I did not notice this when I commented on `peek`. Yes, having such thing for `map` is even worse than for `peek` as we have various `mapToInt`, `mapToLong` and `IntStreamEx.mapToObj` and things like `pairMap` and so on. For consistency I would need to add dozens of methods, which is definitely not an option. See the @nbardiuk comment, that's exactly what should you do.
